### PR TITLE
[dv/common] Add TLUL assert coverage

### PR DESCRIPTION
**Pull request by weicaiyang**
Add following SVA coverage
1. host/device sends back2back req/rst
2. a/d valid drops without ready
3. host uses same source for 2 consecutive items
4. req/rsp content is changed without being accepted

This is interface-level coverage. Will have another PR to add covergroup for TLUL

Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

File: hw/ip/tlul/rtl/tlul_assert.sv [hunk 1]
```diff
@@ -280,6 +280,83 @@ module tlul_assert #(
...
+  endsequence
+
+  // device sends back2back responses
+  sequence b2bRsp_S;
+    d2h.d_valid && h2d.d_ready ##1 d2h.d_valid;
+  endsequence
+
+  // a_valid is dropped without a_ready
+  sequence aValidNotAck_S;
+    h2d.a_valid && !d2h.a_ready ##1 !h2d.a_valid;
```
Comment (cindychip) [inline hw/ip/tlul/rtl/tlul_assert.sv:300]: will there be any cases: `a_ready` is high and then drop to low during `a_valid` is high?

File: hw/ip/tlul/rtl/tlul_assert.sv [hunk 2]
```diff
@@ -280,6 +280,83 @@ module tlul_assert #(
...
+  // d_valid is dropped without a_ready
+  sequence dValidNotAck_S;
+    d2h.d_valid && !h2d.d_ready ##1 !d2h.d_valid;
+  endsequence
+
+  // host uses same source for back2back items
+  sequence b2bSameSource_S;
+    bit [top_pkg::TL_AIW-1:0] pre_source;
+    (h2d.a_valid && d2h.a_ready, pre_source = h2d.a_source) ##1 h2d.a_valid[->1]
+        ##0 pre_source == h2d.a_source;
```
Comment (cindychip) [inline hw/ip/tlul/rtl/tlul_assert.sv:312]: why do we need a ##0 here?

File: hw/ip/tlul/rtl/tlul_assert.sv (hunk 1 above)
Comment (weicaiyang) [reply hw/ip/tlul/rtl/tlul_assert.sv:300]: do you mean below case? if `a_ready` is high, the item is accepted, which isn't related to this assertion.
`h2d.a_valid && d2h.a_ready (this one is accepted) ##1 h2d.a_valid && !d2h.a_ready (this isn't accepted) ##1 !h2d.a_valid (2nd item is dropped);`

Comment (cindychip) [reply hw/ip/tlul/rtl/tlul_assert.sv:300]: thanks for explaning! makes sense

**Review by cindychip — APPROVED** [review]

**Review by tunghoang290780 — APPROVED** [review]
LGTM

File: hw/ip/tlul/rtl/tlul_assert.sv [hunk 3]
```diff
@@ -280,6 +280,92 @@ module tlul_assert #(
...
+    d2h.d_valid && h2d.d_ready ##1 d2h.d_valid;
+  endsequence
+
+  // host sends back2back requests with same address
+  // UVM RAL can't issue this scenario, add this cover to make sure it's tested in some other seq
+  sequence b2bReqWithSameAddr_S;
+    bit [top_pkg::TL_AW-1:0] pre_addr;
+    (h2d.a_valid && d2h.a_ready, pre_addr = h2d.a_address)
+        ##1 h2d.a_valid && pre_addr == h2d.a_address;
+  endsequence
```
Comment (weicaiyang) [inline hw/ip/tlul/rtl/tlul_assert.sv:304]: @cindychip as discussed, I added this, which should cover outstanding requests use the same address

File: hw/ip/tlul/rtl/tlul_assert.sv [hunk 4]
```diff
@@ -280,6 +280,92 @@ module tlul_assert #(
...
+  // d channal content is changed without being accepted
+  `define TLUL_D_CHAN_CONTENT_CHANGED_WO_ACK(NAME) \
+    sequence d_``NAME``ChangedNotAck_S; \
+      int pre; \
+      (d2h.d_valid && !h2d.d_ready, pre = d2h.d_``NAME``) ##1 d2h.d_valid[->1] \
+          ##0 pre != d2h.d_``NAME``; \
+    endsequence \
+    `TLUL_COVER(d_``NAME``ChangedNotAck)
+
+  if (EndpointType == "Host") begin : gen_host_cov
```
Comment (rasmus-madsen) [inline hw/ip/tlul/rtl/tlul_assert.sv:341]: shouldn't these be reversed?
if the dv is acting as a host - we should not check if the device changes response data if no accept.
I mean it is perfectly legal for a device to always keep content stable until it is accepted.

if endpoint type == host means that DV acts as a device I think that merits a comment for clarity.

**Review by rasmus-madsen — APPROVED** [review]
@weicaiyang 
this overall looks very good.

NIT: I for one do not like the term Acknowledge in this scenario - as the ready signal is not an acknowledge.
i.e it can be asserted prior to data, and it can be asserted as a reaction to a valid. but the valid can be deasserted the next clock.

maybe reword to accepted ?

**Review by weicaiyang — COMMENTED** [review]
> @weicaiyang
> this overall looks very good.
> 
> NIT: I for one do not like the term Acknowledge in this scenario - as the ready signal is not an acknowledge.
> i.e it can be asserted prior to data, and it can be asserted as a reaction to a valid. but the valid can be deasserted the next clock.
> 
> maybe reword to accepted ?

As you suggested, changed `ack` to `accepted`. Thanks
